docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+  - Determining which add-ons to install on each Cluster
+  - Resolving the add-on version for the Cluster’s Kubernetes version.
+  - Providing add-on configuration starting from a set of default settings but allowing per Cluster customization.
+  - Waiting for the add-on to be actually running.
+- Some add-ons must be upgraded before, during, or after the cluster upgrade (more specifically the control plane upgrade). This operation brings in some additional requirements such as cleanup the previous version of an add-on.
+- There should be support for out-of band upgrades, e.g. for CVE fix (not linked to the cluster lifecycle).
+- It should be possible to orchestrate pre and post installation steps for each add-on, thus allowing to manage migrations/operations such as:
+  - CPI migration from in-tree to out-of-tree
+  - CoreDNS configuration migration from one version to another
+- Add-on deletion should be orchestrated before Cluster deletion
+
+The current proposal begins to address a subset of these problems by designing a way for Cluster API to orchestrate an external package management tool for add-on management. An external package management tool is expected to provide all the foundational capabilities for add-on management like add-on packages and repositories, templating/configuration, and management of the creation, upgrade, and deletion workflow etc.
+
+In summary, this proposal aims to allow Cluster API and the components designed around it to orchestrate external package managers so that Cluster lifecycle and add-on lifecycle are consistent.
+
+### Goals
+
+- To design a solution for orchestrating Cluster add-ons.
+- To leverage existing package management tools such as Helm for all the foundational capabilities of add-on management, i.e. add-on packages/repository, templating/configuration, add-on creation, upgrade and deletion etc. 
+- To make add-on management in Cluster API modular and pluggable, and to make it simple for developers to build a Cluster API Add-on Provider based on any package management tool, just like with infrastructure and bootstrap providers.
+
+### Non goals
+
+- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
+- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on, i.e. Deployments, Services, ServiceAccounts. Cluster API should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
+- To expect users to use a specific package management tool.
+- To implement a solution for installing add-ons on the management cluster itself.
+
+### Future work
+
+- Handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version.
```

Comment:
@Jont828  am I wrong or this is already covered by the current implementation

Reply:
We don't really have a way to trigger a Helm upgrade when a Cluster is upgraded. You can set a new chart version in HelmChartProxy and it can update on the next reconciliation though.